File: net-event-bus/examples/observe.go
// What the drop counters actually tell you (Go).
//
// Run: go run observe.go
//
// THE POINT OF THIS FILE, AND IT IS NOT WHAT YOU EXPECT.
//
// The usual advice — "backpressure is silent, so watch EventsDropped" — is
// wrong for the one mode where it matters most, and that mode is the default.
//
// Both counters sit at the producer boundary: they record what the bus
// accepted or refused from you, not what survived to an adapter.
//
//   - DropOldest (the default) evicts to make room, so IngestRaw always
//     succeeds and EventsDropped never moves. Events are lost and nothing
//     in-process says so.
//   - DropNewest and FailProducer refuse the producer instead, so the ingest
//     call and EventsDropped both see it.
//
// The exact counts vary between runs and between bindings — the drain worker
// races the producer. The shape is what matters, not the numbers.
//
// TWO GO NAMING TRAPS, both compile errors rather than bugs:
//
//   - The struct does NOT use the C header's snake_case names. It is
//     EventsIngested / EventsDropped. The JSON tags carry snake_case; the Go
//     fields do not.
//   - The batch counter is spelled BatchesDispathed — missing the second 'c'.
//     That typo is the actual field name, so stats.BatchesDispatched does not
//     compile.
package main

import (
	"fmt"
	"log"

	"github.com/ai-2070/net/go"
)

func measure(mode string) {
	// RingBufferCapacity must be a power of two and >= 1024 — both enforced at
	// construction, and neither is caught by the compiler.
	bus, err := net.New(&net.Config{
		NumShards:          1,
		RingBufferCapacity: 1024,
		BackpressureMode:   mode,
	})
	if err != nil {
		log.Fatal(err)
	}
	defer bus.Shutdown()

	refused := 0
	for i := 0; i < 20000; i++ {
		if err := bus.IngestRaw(`{"seq":1}`); err != nil {
			refused++
		}
	}

	// Stats returns (*Stats, error) in Go — the other bindings return the
	// struct directly.
	stats, err := bus.Stats()
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("%-14s ingested=%-6d dropped=%-6d batches=%-6d ingest_errors=%d\n",
		mode, stats.EventsIngested, stats.EventsDropped,
		stats.BatchesDispathed, // sic — see the header comment
		refused)
}

func main() {
	fmt.Println("20,000 events into a 1024-slot buffer, nothing consuming:")
	fmt.Println()
	// Either casing parses: "DropOldest" and "drop_oldest" are the same thing.
	// An unrecognised value is rejected outright rather than defaulting.
	for _, mode := range []string{"DropOldest", "DropNewest", "FailProducer"} {
		measure(mode)
	}
	fmt.Println()
	fmt.Println("DropOldest loses events silently — and it is the default.")
	fmt.Println("The other two refuse the producer, so ingest and the counter agree.")
}

File: net-event-bus/examples/observe.rs
//! What the drop counters actually tell you (Rust).
//!
//! Drop this into your crate's `examples/` directory and run:
//!   cargo run --example observe
//!
//! Cargo.toml: same dependencies as `hello.rs`.
//!
//! THE POINT OF THIS FILE, AND IT IS NOT WHAT YOU EXPECT.
//!
//! The usual advice — "backpressure is silent, so watch `events_dropped`" — is
//! wrong for the one mode where it matters most, and that mode is the default.
//!
//! Both counters sit at the **producer boundary**: they record what the bus
//! accepted or refused from you, not what survived to an adapter.
//!
//!   - `DropOldest` (the default) evicts to make room, so the producer always
//!     succeeds and `events_dropped` never moves. Events are lost and nothing
//!     in-process says so.
//!   - `DropNewest` and `FailProducer` refuse the producer instead, so the
//!     `emit` call *and* `events_dropped` both see it.
//!
//! The exact counts vary between runs and between bindings: the drain worker
//! races the producer, so how full the buffer gets is a timing question. The
//! shape is what matters, not the numbers.

use net_sdk::{Backpressure, Net};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
struct Tick {
    seq: u64,
}

/// Push far past capacity with nothing consuming, and report the counters.
async fn measure(label: &str, mode: Backpressure) -> net_sdk::error::Result<()> {
    let node = Net::builder()
        .shards(1)
        .buffer_capacity(1024) // power of two, and >= 1024 — both enforced
        .backpressure(mode)
        .memory()
        .build()
        .await?;

    let mut refused = 0u32;
    for seq in 0..20_000 {
        if node.emit(&Tick { seq }).is_err() {
            refused += 1;
        }
    }

    let s = node.stats();
    println!(
        "{label:<13} ingested={:<6} dropped={:<6} emit_errors={}",
        s.events_ingested, s.events_dropped, refused
    );

    node.shutdown().await
}

#[tokio::main(flavor = "current_thread")]
async fn main() -> net_sdk::error::Result<()> {
    println!("20,000 events into a 1024-slot buffer, nothing consuming:\n");
    measure("DropOldest", Backpressure::DropOldest).await?;
    measure("DropNewest", Backpressure::DropNewest).await?;
    measure("FailProducer", Backpressure::FailProducer).await?;
    println!("\nDropOldest loses events silently — and it is the default.");
    println!("The other two refuse the producer, so emit() and the counter agree.");
    Ok(())
}
